rc/lsp.kak

```diff
@@ -228,18 +228,157 @@ define-command lsp-hover -docstring "Request hover info for the main cursor posi
+    lsp-did-change-and-then "lsp-goto-next-symbol-request %arg{@}"
+}
+
+define-command -hidden lsp-goto-next-symbol-request -params 0..1 -shell-script-candidates %opt{lsp_symbol_kind_completion} \
+    -docstring "lsp-goto-next-symbol-request [<symbol-kind>]: Goto to the next symbol of type <symbol-kind> in the document. If <symbol-kind> is not given, it means _any_ symbol" %{
+    nop %sh{ (printf '
+session   = "%s"
+client    = "%s"
+buffile   = "%s"
+filetype  = "%s"
+version   = %d
+method    = "kak-lsp/next-or-previous-symbol"
+[params]
+position.line      = %d
+position.column    = %d
+symbolKind = "%s"
+searchNext = true
+' "${kak_session}" "${kak_client}" "${kak_buffile}" "${kak_opt_filetype}" "${kak_timestamp}" ${kak_cursor_line} ${kak_cursor_column} "$1" | eval ${kak_opt_lsp_cmd} --request) > /dev/null 2>&1 < /dev/null & }
+}
+
+# Frequently used, provided for convenience
+define-command lsp-goto-next-function -params 0 \
+    -docstring "lsp-goto-next-function: Goto to the next function in the document" %{
+    lsp-goto-next-symbol Function
+}
+
+# Frequently used, provided for convenience
+define-command lsp-goto-previous-function -params 0 \
+    -docstring "lsp-goto-previous-function: Goto to the previous function in the document" %{
+    lsp-goto-previous-symbol Function
+}
+
+# Frequently used, provided for convenience
+define-command lsp-goto-next-method -params 0 \
+    -docstring "lsp-goto-next-method: Goto to the next method in the document" %{
+    lsp-goto-next-symbol Method
+}
+
+# Frequently used, provided for convenience
+define-command lsp-goto-previous-method -params 0 \
+    -docstring "lsp-goto-previous-method: Goto to the previous method in the document" %{
+    lsp-goto-previous-symbol Method
+}
+
+define-command lsp-hover-previous-symbol -params 0..1 -shell-script-candidates %opt{lsp_symbol_kind_completion} \
+    -docstring "lsp-hover-previous-symbol [<symbol-kind>]: Show hover of the previous symbol of type <symbol-kind> in the document. If <symbol-kind> is not given, it means _any_ symbol. The editor does _not_ actually navigate to that symbol. This command is useful to peek at the symbol hover information" %{
+    lsp-did-change-and-then "lsp-hover-previous-symbol-request %arg{@}"
+}
+
+define-command -hidden lsp-hover-previous-symbol-request -params 0..1 -shell-script-candidates %opt{lsp_symbol_kind_completion} \
+    -docstring "lsp-hover-previous-symbol-request [<symbol-kind>]: Show hover of the previous symbol of type <symbol-kind> in the document. If <symbol-kind> is not given, it means _any_ symbol. The editor does _not_ actually navigate to that symbol. This command is useful to peek at the symbol hover information" %{
+    nop %sh{ (printf '
+session   = "%s"
+client    = "%s"
+buffile   = "%s"
+filetype  = "%s"
+version   = %d
+method    = "kak-lsp/next-or-previous-symbol"
```

Comment:
We should probably just have one hidden `lsp-next-or-previous-symbol <next> <hover> <symbol>`.
We could even make autocompletion work with `-shell-script-completion` but since it's a hidden command it doesn't matter.

Reply:
This was a good suggestion and I have created `lsp-next-or-previous-symbol` and everything is funneled through it.